src/actions/send.js

```diff
@@ -0,0 +1,40 @@
+import { createActions } from 'redux-actions';
+
+import { showAlert } from '../utils/alerts';
+import { wallet } from '../utils/wallet';
+
+export const transfer = ({ contractName, amount, memo, receiverId, isStorageBalanceAvailable }) => async (dispatch, gesState) => {
+    if (contractName && !isStorageBalanceAvailable) {
+        await dispatch(send.transfer.storageDeposit(contractName, receiverId));
+    }
+
+    const { transaction, status } = await dispatch(send.transfer[contractName ? 'TOKENS' : 'NEAR'](contractName, amount, memo, receiverId));
+
+    if (status?.SuccessValue) {
+        dispatch(send.setTxStatus(transaction.hash, 'success'));
+    }
+};
+
+export const { send } = createActions({
+    SEND: {
+        TRANSFER: {
+            NEAR: [
+                wallet.sendMoney.bind(wallet),
+                () => showAlert({ onlyError: true })
+            ],
+            TOKENS: [
+                () => {},
+                () => showAlert({ onlyError: true })
+            ],
+            STORAGE_DEPOSIT: [
+                () => {},
+                () => showAlert({ onlyError: true })
+            ],
+        },
+        IS_STORAGE_BALANCE_AVAILABLE: [
```

Comment:
It's a bit weird that this is named like it is a boolean value.  Would this be more like e.g. `SET_STORAGE_BALANCE_IS_AVAILABLE`, or `LOAD_STORAGE_BALANCE_IS_AVAILABLE`?

Reply:
This action will call `isStorageBalanceAvailable` method from `FungibleTokens` class. 

The general idea that stays behind it, is that I would like to omit direct calling methods in components - ideal will be if the component should only be aware of redux actions and operates only with redux actions - also, all logic should be inside redux actions, so the component will do simple dispatch and that's it. 

Of course in this particular situation, `isStorageBalanceAvailable` action will not make any changes in the redux state, but that's because we are not storing any state for send flow. But still, I think that it should be done this way because it will make components much more clear about what they do, and move all logic into actions.

The additional benefit with making all functionality with redux actions would be that we could then wrap more logic into this, like for example mixpanel - at this moment mixpanel tracking is inside components, which is not ideal, I think it could be done automatically with redux.

Not sure if I explain my idea behind it properly :)

Reply:
One important thing here :) all actions are handled by `status` reducer, we are storing the result of all actions and we can show Global Alert / Local Alert automatically if particular action fails or succeded.

> I also don't think that it makes it much more clear what the component is doing; it makes me think that the component is updating Redux state, but in actuality the component is just running a method on the Wallet object. 

We can react automatically if an error appears - we don't need to handle every method individually.

> The way this is now, in order for me to know whether or not a particular call has a Redux side-effect, I need to search the entire codebase for the SNAKE_CASE version of the action name; by finding no occurrences outside of the actions file of that string, I can assume that we don't use it to mutate redux. 

The component is dispatching thunk action, hi-level operations that this action does are described by its name - you just jump to this thunk action, and everything that is happening is in there (instead of component). So we use getState there to take all necessary states, and we are dispatching all actions that are needed. Actions are composed of objects, like `send.transfer.nep141` or `send.isStorageDepositAvailable` so it's clear what they are doing - you can also jump quickly to the method, to see details there.

The benefit is that we are moving all logic into thunk actions from the component. Also if like, in this case, we are composing FungibleTokens class, to be like low-level functions only, the thunk actions are the level above that uses these actions/methods to do all things.

> Wouldn't it be nice if I could immediately tell that something doesn't mutate redux at the call-site because it isn't a Redux action creator?

All actions are mutating the state, in this case `status`.

I think that the redux should be the center of our application and the only source of truth, the application should be rendered based on the redux state. We should store everything that is happening in the application. For example, if some request fails, we know it by checking redux.

I mean if we want classes to be low-level functions only, and we don't want to have complex logic in components directly (I think we don't because it's complicating stuff, the component needs to have all redux state to perform a sometimes complicated set of actions), I think the good idea is to handle everything in thunk actions - we have access to redux state there, and we can dispatch everything necessary - component only dispatches one action. If we don't want to use thunk actions, we should implement another level of logic, to handle everything that is happening between class methods and components.

I mean that is my idea, how to handle it. If you have other concepts here, no problem at all, I'm open to new ideas, just let me know how would you like to handle it :)

Reply:
Thanks for the thoughtful response -- I'll write up an issue about this topic so we can address it directly rather as part of this PR discussion :)